[WIP] Linux support

qyot27 opened the pull request:
**Currently incomplete; this pull request is for discussion and bug testing**

The following gets us more or less over the hurdle of all Windows-specific sections being sectioned off so we know which parts need porting (or in some cases, not - a lot of `main.cpp` looks like it's just for Video for Windows or DirectShow).  To find these, look for the AVS_WINDOWS defines.  For some of the non-Win32 parts of the ifdefs, I did copy over some things from AvxSynth where the code seemed adequate (acknowledging that it may have to be heavily refactored or rewritten in the future in order to bring it up to par with where AviSynth+ is or to even unify them into a cross-platform version), and the types used have been mostly switched from MSVC types to C/C++ std types.  It still has vital parts of the code missing on the Linux side, though.

Exprfilter doesn't build on Linux, and neither does Text Overlay (the latter because it uses Win32-specific APIs).  The lack of Text Overlay actually is the root cause of a lot of other issues I had getting things to compile, so none of the text-related stuff like Info, Version, or Subtitle will work unless we get Text Overlay working.  The logging ability also had to be disabled in Linux builds.

With that in mind, though, this does result in a libavisynth.so library being compiled, and I've been able to test [against a patch of FFmpeg's libavformat AviSynth demuxer that I modifed to look for AviSynth+ on Linux instead of AvxSynth](https://github.com/qyot27/FFmpeg/tree/avsplus_linux).  It segfaults, but the output of `LD_DEBUG=libs ./ffmpeg -i test.avs` no longer reports that symbols are missing, like it was before.

I was thinking that, if we can get Colorbars() working and producing output, that's a good first step and the bare minimum criteria for merging what we have right now (hopefully as part of 3.5).

Linux testing was done on both Ubuntu 19.10, and on WSL with Ubuntu 19.10 loaded into it.  Off the top of my head, dependencies should be just `build-essential`, `cmake` and `git`, and optionally `checkinstall`.  Building the library should be as simple as:
```
cd AviSynthPlus
mkdir avisynth-build
cmake ../ -DCMAKE_CXX_FLAGS="-fpermissive"
make
    sudo checkinstall --pkgname=avisynth --pkgversion="$(grep -r \
    Version avs_core/avisynth.pc | cut -f2 -d " ")-$(date --rfc-3339=date | \
    sed 's/-//g')-git" --backup=no --deldoc=yes --delspec=yes --deldesc=yes \
    --strip=yes --fstrans=no --default
```
`-fpermissive` is still required to get it to build for Linux, and particularly, I suspect that the issues around BufferPool and ThreadPool that stop cold without `-fpermissive` might be the biggest thing standing in the way of getting basic output with FFmpeg.  I may be wrong, though.

pinterf commented:
Nice work. There was a suprisingly huge amount of old-style types such as __int64 instead of int64_t. 
As for Expr filter: it has a properly working C-only branch, it's only the jitasm part that has to be disabled first if it has any build problems. In the base code from VapourSynth jit was working for linux as well, only the non-x86 processors were omitted there.

pinterf commented:
I think text overlay will never work the same way as it does now in Windows, only by using an external plugin. How about revive info.c/info.h, which is a debug-like output using monospace fonts. Its only usage now is from histogram. Stainlesss' [DDigits](https://forum.doom9.org/showthread.php?p=1462089) is based on that and a bit smarter but is still for simple text output.

qyot27 commented:
I was thinking about whether libass could be used to fill that particular need, even if we end up having to split text overlay into a _win32 variant and a _unix variant (with libass used by _unix).  While it's most notably a subtitling library, it is used in mpv to do the rendering of the OSD.

The way AvxSynth did it was by integrating the Pango and Cairo libraries, but I'm not very keen on that as doing that also pulls in GLib as a dependency (which is huge, and if you're not using the repository version, compiling GLib is a nightmare) and while I'm not sure that the text rendering library choice was the reason for it, the output of Version() under AvxSynth is very different from AviSynth.

pinterf commented:
See [Subtext](https://forum.doom9.org/showthread.php?t=173748) here

pinterf commented:
Two questions in case I can help you in developing linux things.
1.) I've successfully installed WSL 18.04LTS and upgraded to 19.10. Cloned Avisynth repository.
Now how can I apply your existing pull requests without doing that for the master github repository?
I'm not confident with the necessary commands. Or should I use your repo linux branch?
2.) Then I have to clone your modified ffmpeg repository then build it somehow, I'd need some hints as well. Thanks.

pinterf commented:
A minor note: `__int64` became `long long int` in focus.cpp (and its counterparts), at all other places it was changed to `int64_t`

qyot27 commented:
You mean how to push the commits without pushing directly to the linux branch here?  Checkout the linux branch locally, make any changes, push the linux branch up to your own repo, and then open a pull request against the linux branch here so it goes from `pinterf/linux` to `AviSynth/linux`.  That's probably the safest option.

With FFmpeg, you can probably get it going by first editing /etc/sources.list and uncommenting the source repositories so build-dep knows what to do, then:
```
sudo apt-get build-dep ffmpeg
```
That will grab all the repository prerequisites for building ffmpeg.  As far as the patch in my repo goes, 
```
git clone git://github.com/qyot27/FFmpeg.git
cd FFmpeg
git checkout avsplus_linux
 ./configure --prefix=$HOME/ffavx_build --enable-gpl --enable-version3 \
    --disable-doc --disable-stripping --enable-avisynth --disable-optimizations \
    --extra-cflags="-I/usr/local/include/avisynth" && \
make -j$(nproc)
make install
```

IIRC, the `long long int` in `focus.cpp` rather than `int64_t` was due to a build error if I tried using `int64_t`.  Occasionally (but not often), that did end up happening.

pinterf commented:
Oops, I did not recognize that linux branch exists here and not only at your repo.

pinterf commented:
I'm stuck
> With FFmpeg, you can probably get it going by first editing /etc/sources.list and uncommenting the source repositories so build-dep knows what to do, then:
> 
> ```
> sudo apt-get build-dep ffmpeg
> ```
Either I uncomment everything or not in /etc/apt/sources.list build-dep reports:
```
Reading package lists... Done
E: You must put some 'source' URIs in your sources.list

```

qyot27 commented:
Did you run `sudo apt-get update` after uncommenting the stuff in sources.list?  For what it's worth, this is what mine looks like:
```
# See http://help.ubuntu.com/community/UpgradeNotes for how to upgrade to
# newer versions of the distribution.
deb http://archive.ubuntu.com/ubuntu/ eoan main restricted
deb-src http://archive.ubuntu.com/ubuntu/ eoan main restricted

## Major bug fix updates produced after the final release of the
## distribution.
deb http://archive.ubuntu.com/ubuntu/ eoan-updates main restricted
deb-src http://archive.ubuntu.com/ubuntu/ eoan-updates main restricted

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team. Also, please note that software in universe WILL NOT receive any
## review or updates from the Ubuntu security team.
deb http://archive.ubuntu.com/ubuntu/ eoan universe
deb-src http://archive.ubuntu.com/ubuntu/ eoan universe
deb http://archive.ubuntu.com/ubuntu/ eoan-updates universe
deb-src http://archive.ubuntu.com/ubuntu/ eoan-updates universe

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team, and may not be under a free licence. Please satisfy yourself as to
## your rights to use the software. Also, please note that software in
## multiverse WILL NOT receive any review or updates from the Ubuntu
## security team.
deb http://archive.ubuntu.com/ubuntu/ eoan multiverse
deb-src http://archive.ubuntu.com/ubuntu/ eoan multiverse
deb http://archive.ubuntu.com/ubuntu/ eoan-updates multiverse
deb-src http://archive.ubuntu.com/ubuntu/ eoan-updates multiverse

## N.B. software from this repository may not have been tested as
## extensively as that contained in the main release, although it includes
## newer versions of some applications which may provide useful features.
## Also, please note that software in backports WILL NOT receive any review
## or updates from the Ubuntu security team.
deb http://archive.ubuntu.com/ubuntu/ eoan-backports main restricted universe multiverse
deb-src http://archive.ubuntu.com/ubuntu/ eoan-backports main restricted universe multiverse

## Uncomment the following two lines to add software from Canonical's
## 'partner' repository.
## This software is not part of Ubuntu, but is offered by Canonical and the
## respective vendors as a service to Ubuntu users.
# deb http://archive.canonical.com/ubuntu xenial partner
# deb-src http://archive.canonical.com/ubuntu xenial partner

deb http://security.ubuntu.com/ubuntu/ eoan-security main restricted
deb-src http://security.ubuntu.com/ubuntu/ eoan-security main restricted
deb http://security.ubuntu.com/ubuntu/ eoan-security universe
deb-src http://security.ubuntu.com/ubuntu/ eoan-security universe
deb http://security.ubuntu.com/ubuntu/ eoan-security multiverse
deb-src http://security.ubuntu.com/ubuntu/ eoan-security multiverse
```

pinterf commented:
> Did you run `sudo apt-get update` after uncommenting the stuff in sources.list? For what it's worth, this is what mine looks like:

Much better, thanks, I'm slowly getting a grip on it

pinterf commented:
> ./configure --prefix=$HOME/ffavx_build --enable-gpl --enable-version3 \
>     --disable-doc --disable-stripping --enable-avisynth --disable-optimizations \
>     --extra-cflags="-I/usr/local/include/avisynth" && \
> make -j$(nproc)

Turned out that this `--extra-cflags="-I/usr/local/include/avisynth"` points to nothing on my system, resulting in
```
libavformat/avisynth.c:44:12: fatal error: avisynth/avisynth_c.h: No such file or directory
   44 |   #include <avisynth/avisynth_c.h>
      |            ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [ffbuild/common.mak:59: libavformat/avisynth.o] Error 1
make: *** Waiting for unfinished jobs....

```
libavformat/avisynth.c was modified to use the compat directory for non _WIN32 and undefined AVSC_NO_DECLSPEC:
/* Enable function pointer definitions for runtime loading. */
#define AVSC_NO_DECLSPEC

```
/* Platform-specific directives for AviSynth vs AvxSynth. */
#ifdef _WIN32
  #include "compat/w32dlfcn.h"
  #undef EXTERN_C
  #include "compat/avisynth/avisynth_c.h"
  #define AVISYNTH_LIB "avisynth"
  #define USING_AVISYNTH
#else
  #include <dlfcn.h>
//  #include "compat/avisynth/avxsynth_c.h"
//  #include <avisynth/avisynth_c.h>
  #include "compat/avisynth/avisynth_c.h"
  #undef AVSC_NO_DECLSPEC
  #define AVISYNTH_NAME "libavisynth"
  #define AVISYNTH_LIB AVISYNTH_NAME SLIBSUF
#endif 

```
Then modded capi.h a bit and the compilation was finally successful.
Have you commited everything there?

pinterf commented:
> Off the top of my head, dependencies should be just `build-essential`, `cmake` and `git`. Building the library should be as simple as:

Just as a memo: `checkinstall` is also a dependancy

pinterf commented:
O.K. ffmpeg and avisynth is make installed, I'm sure I missed something, made a Test directory with a simple ColorBars in it, and copied ffmpeg here, but I'm getting unknown error
./ffmpeg -loglevel debug -i test.avs test.mkv
```
ffmpeg version N-96439-ge2d498f6d7 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.2.1-9ubuntu2)
  configuration: --prefix=/home/pinterf/ffavx_build --enable-gpl --enable-version3 --disable-doc --disable-stripping --enable-avisynth --disable-optimizations --extra-cflags=-I/usr/local/include/avisynth
  libavutil      56. 38.100 / 56. 38.100
  libavcodec     58. 66.100 / 58. 66.100
  libavformat    58. 35.104 / 58. 35.104
  libavdevice    58.  9.103 / 58.  9.103
  libavfilter     7. 71.100 /  7. 71.100
  libswscale      5.  6.100 /  5.  6.100
  libswresample   3.  6.100 /  3.  6.100
  libpostproc    55.  6.100 / 55.  6.100
Splitting the commandline.
Reading option '-loglevel' ... matched as option 'loglevel' (set logging level) with argument 'debug'.
Reading option '-i' ... matched as input url with argument 'test.avs'.
Reading option 'test.mkv' ... matched as output url.
Finished splitting the commandline.
Parsing a group of options: global .
Applying option loglevel (set logging level) with argument debug.
Successfully parsed a group of options.
Parsing a group of options: input url test.avs.
Successfully parsed a group of options.
Opening an input file: test.avs.
[NULL @ 0x7fffddda8280] Opening 'test.avs' for reading
[file @ 0x7fffddda8e00] Setting default whitelist 'file,crypto,data'
[avisynth @ 0x7fffddda8280] Format avisynth probed with size=2048 and score=50
[AVIOContext @ 0x7fffdddb1280] Statistics: 12 bytes read, 0 seeks
test.avs: Unknown error occurred
```

qyot27 commented:
Keeping the avisynth/ compat headers around in FFmpeg tends to get somewhat heated whenever it gets brought up on the mailing list, so I wanted to make sure it could also use the headers that AviSynth+ installs to the system.

The changes to avisynth_c.h and capi.h dealing with FFmpeg integration are part of this pull request, in these commits:
https://github.com/AviSynth/AviSynthPlus/pull/120/commits/d8a23cab7b2dc340671d91a07c70b1cb546299f2

https://github.com/AviSynth/AviSynthPlus/pull/120/commits/713b6d797822621e2ebffe5bab9245287f7c6105

`checkinstall` is actually optional, since you could have done `sudo make install` followed by `sudo ldconfig`.  What `checkinstall` does (apart from also automatically running ldconfig to make the system aware of the newly-installed library) is construct a deb package and log it into the package management system so that the user can keep track of it that way, or more easily remove/uninstall the package if need be.

pinterf commented:
Yep, avisynth_c in ffmpeg is a bit hacked. It was a pain to debug the crashes, putting env->ThrowError("blabla I am here or there") positioned away line by line.
Anyway. We have a working ColorBars() through ffmpeg :)
Sources need heavy polishing, maybe it's best to zip them together and send you tomorrow.

pinterf commented:
Here are the modified files, ffmpeg, avisynth, and the gigatest script with Colorbars :)
Some have a lot of debug print, mainly in ffmpeg.
https://drive.google.com/open?id=1oCA-hscfmpwQcGhgITnYWchwxJ-Mfh83

qyot27 commented:
Oh wow, the changes needed were a lot smaller than I thought.  The block in `parser/script.cpp` was actually the worst of it; the block in `avisynth.cpp` was able to be greatly simplified, and no actual changes were really needed for FFmpeg, since yeah, most of that had been done already for the system headers.  I'll fixup the changes in a commit under your name and add it to the pull request.

Side-note: Linux (well, pretty much *every* OS apart from Windows) uses UTF-8 as the native codepage, obviating the need for the kinds of special tricks we need to handle that sort of thing on Windows.  I left the FFmpeg block for Unicode handling alone, and FFmpeg and libavisynth.so was able to not only open `ウィキペディア日本語版.avs` but it also could open a script that just did `Import("ウィキペディア日本語版.avs")`.  This was on actual Ubuntu, not WSL.  I wouldn't be surprised if WSL is a little weird on this particular issue.  It does mean I really need to clean up the patch for FFmpeg to actually not do the weird jumping around with `USING_AVISYNTH` ifdefs, as we'll no longer need them at all - the few places it's necessary can revert to simple `_WIN32` checks.

One thing that this did expose is that the cpuid stuff **really** needs to be fixed up for Linux.  ConvertTo causes crashes because the cpuid isn't being handled correctly (it's also probably causing a segfault in AddBorders, but that only happens when exiting, as the output has the borders added).  So long as you generate Colorbars or BlankClip, etc. in the pixel_type you want, it's okay with any of them.  But for the moment we can't convert in-script.

You can also do `sudo apt-get install libsdl2-dev`, and then rebuild FFmpeg, as that will generate ffplay and the script can be previewed directly (on WSL, you'll also need to install/setup Xming like [this article](https://virtualizationreview.com/articles/2017/02/08/graphical-programs-on-windows-subsystem-on-linux.aspx) demonstrates; you can then add the `export DISPLAY=:0` to WSL's .bashrc file so it always hooks in on startup...just make sure that Xming is started first before launching WSL).

magiblot commented:
I have ran AviSynth with pinterf's changes and noted the following:
* `stdc++fs` needs to be added as link library in `avs_core/CMakeLists.txt`
* `__cpuid` is not well defined in `avs_core/core/cpuid.cpp`. The following seems to be more accurate:
```cpp
#include <cpuid.h>
#undef __cpuid
static inline void __cpuid(int cpuinfo[4], int id) {
    unsigned int *i = reinterpret_cast<unsigned int *>(cpuinfo);
    __get_cpuid(id, &i[0], &i[1], &i[2], &i[3]);
}
```
Although for cases like this one, I would find it easier to write an interface with a different implementation for every platform instead of placing ifdefs everywhere.
* StackVertical/StackHorizontal/Animate crash when the script ends or is interrupted:
```
BlankClip()
StackHorizontal(last, last)
```
The cause is `vfb` being `0` in `VideoFrame::Release()`:
```
Thread 1 "ffmpeg" received signal SIGSEGV, Segmentation fault.
0x00007ffffc07568d in VideoFrame::Release (this=0xa4d33f0) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/interface.cpp:481
481         InterlockedDecrement(&_vfb->refcount);
(gdb) bt
#0  0x00007ffffc07568d in VideoFrame::Release (this=0xa4d33f0) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/interface.cpp:481
#1  0x00007ffffc075aae in VideoFrame::DESTRUCTOR (this=0xa4d33f0) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/interface.cpp:567
#2  0x00007ffffc075a92 in VideoFrame::~VideoFrame (this=0xa4d33f0, __in_chrg=<optimized out>) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/interface.cpp:566
#3  0x00007ffffc0560b1 in ScriptEnvironment::~ScriptEnvironment (this=0xa4be8b0, __in_chrg=<optimized out>) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/avisynth.cpp:1147
#4  0x00007ffffc056224 in ScriptEnvironment::~ScriptEnvironment (this=0xa4be8b0, __in_chrg=<optimized out>) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/avisynth.cpp:1174
#5  0x00007ffffc05c798 in ScriptEnvironment::DeleteScriptEnvironment (this=0xa4be8b0) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/avisynth.cpp:3170
#6  0x00007ffffc072390 in avs_delete_script_environment (e=0xa4be890) at /home/magiblot/Documents/AviSynthPlus/avs_core/core/avisynth_c.cpp:887
#7  0x00000000084ddb94 in avisynth_context_destroy (avs=0xa4a7100) at libavformat/avisynth.c:215
#8  0x00000000084dea39 in avisynth_read_close (s=0xa4a6200) at libavformat/avisynth.c:858
#9  0x0000000008658b0b in avformat_close_input (ps=0xa4d2880) at libavformat/utils.c:4492
#10 0x000000000811a651 in ffmpeg_cleanup (ret=0) at fftools/ffmpeg.c:590
#11 0x00000000081126c8 in exit_program (ret=0) at fftools/cmdutils.c:139
#12 0x000000000812ab06 in main (argc=4, argv=0x7ffffffee268) at fftools/ffmpeg.c:4901
(gdb) print vfb
$3 = (VideoFrameBuffer *) 0x0
```

qyot27 commented:
@magiblot 
That's GCC 8, isn't it?  `stdc++fs` has [apparently already been folded back into `stdc++` in GCC 9](https://gitlab.kitware.com/cmake/cmake/issues/17834), which is why this wasn't immediately obvious.  Although it does raise the question of whether Clang needs `c++fs` linked in.

And yeah, cpuid needs an overhaul, the messages about BufferPool and ThreadPool also concern me (and those are fatal if not using `fpermissive`, so I'm sure for good reason - if that ties into the state `vfb` is left in, that might explain the filters crashing on exit).

I've also now cleaned up and updated the FFmpeg patchset to what is probably its final version (depending on how updating the compat headers goes, and whether the consensus on the FFmpeg-devel ML is that there should be a non-fatal warning message alerting users to AvxSynth's future removal).  The branch has been force-pushed to my FFmpeg repo, so take the necessary precautions before trying to update there.

pinterf commented:
>     * StackVertical/StackHorizontal/Animate crash when the script ends or is interrupted:
> 
> 
> ```
> BlankClip()
> StackHorizontal(last, last)
> ```
> 
> The cause is `vfb` being `0` in `VideoFrame::Release()`:

I have a faint clue from past years that I saw similar problems when using C interface, double clip release or other resource free-up occurs somewhere in ffmpeg? I have to look at what atexit is doing.

pinterf commented:
> Side-note: Linux (well, pretty much _every_ OS apart from Windows) uses UTF-8 as the native codepage,

Yes, I also noticed that we don't need any magic at either side. utf8 comes and goes, it should work transparently.

pinterf commented:
> I'll fixup the changes in a commit under your name and add it to the pull request.

Thanks, I didn't have time for that (plus I'm still not too familiar with linux and its console-driven git usage and development so I can do only what you and stackoverflow suggest me), and I left a little mess there, but I was way too curious and had to test things quickly and use the comment out or copy paste and see what happens method, so I shared you the code as-is.

magiblot commented:
Yes, I was using GCC 8.
Why is it necessary to remove AvxSynth support from FFmpeg, though?

pinterf commented:
> messages about BufferPool and ThreadPool also concern me (and those are fatal if not using `fpermissive`, so I'm sure for good reason - if that ties into the state `vfb` is left in, that might explain the filters crashing on exit).

They are probably just a bad naming choice that the very names of the class types are chosen to be a field name.

qyot27 commented:
>Why is it necessary to remove AvxSynth support from FFmpeg, though?

AvxSynth is basically AviSynth 2.5, which FFmpeg doesn't support on the Windows side either.  That means there's several places in the demuxer with special casing just to support it, because there were parts of the C API that changed.  There's also name conflicts and which headers to include (because AviSynth+'s headers will not work as a substitute to load libavxsynth).

pinterf commented:
> ```
> BlankClip()
> StackHorizontal(last, last)
> ```
> 
> 
> The cause is `vfb` being `0` in `VideoFrame::Release()`:

It was a dirty one.
Fixed in master: e4bf883e9e64be3bd3f77611eed766c8ffa4a08b

qyot27 commented:
The linux branch has been rebased against master (for the vfb fix) and force-pushed.

Okay, testing both @pinterf's and @magiblot's changes allows Convert to work and there seem to be no more crashes/segfaults on exit with a simple Colorbars.  Even SetFilterMTMode and Prefetch seem to work - `ColorbarsHD().ConvertBits(16)` is ~94fps on my J3455, setting the MT Mode to MT_NICE_FILTER and Prefetch(4) results in ~194fps.

But unfortunately, on the complex test script pinterf posted in the 7z archive above, it just segfaults:
```
(gdb) r -i test_complex.avs 
Starting program: /home/qyot27/ffavx_build/bin/ffmpeg -i test_complex.avs
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
ffmpeg version N-96735-g9356fecae8 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.2.1-9ubuntu2)
  configuration: --prefix=/home/qyot27/ffavx_build --enable-gpl --enable-version3 --disable-doc --disable-stripping --enable-avisynth --disable-optimizations --extra-cflags=-I/usr/local/include/avisynth
  libavutil      56. 41.100 / 56. 41.100
  libavcodec     58. 68.102 / 58. 68.102
  libavformat    58. 38.100 / 58. 38.100
  libavdevice    58.  9.103 / 58.  9.103
  libavfilter     7. 75.100 /  7. 75.100
  libswscale      5.  6.100 /  5.  6.100
  libswresample   3.  6.100 /  3.  6.100
  libpostproc    55.  6.100 / 55.  6.100
[New Thread 0x7ffff613d700 (LWP 7062)]
[New Thread 0x7ffff593c700 (LWP 7063)]
[New Thread 0x7ffff513b700 (LWP 7064)]
[New Thread 0x7ffff493a700 (LWP 7065)]

Thread 1 "ffmpeg" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff6672f7c in Exprfilter::GetFrame(int, IScriptEnvironment*) ()
   from /usr/local/lib/libavisynth.so
#2  0x00007ffff65e74a1 in Cache::GetFrame(int, IScriptEnvironment*) ()
   from /usr/local/lib/libavisynth.so
#3  0x00007ffff65e5560 in avs_get_frame () from /usr/local/lib/libavisynth.so
#4  0x0000555555a62cd2 in avisynth_read_packet_video (s=0x555557b4d600, 
    pkt=0x7fffffffd440, discard=0) at libavformat/avisynth.c:671
#5  0x0000555555a63247 in avisynth_read_packet (s=0x555557b4d600, 
    pkt=0x7fffffffd440) at libavformat/avisynth.c:809
#6  0x0000555555be7ea1 in ff_read_packet (s=0x555557b4d600, pkt=0x7fffffffd440)
    at libavformat/utils.c:857
#7  0x0000555555beb0dc in read_frame_internal (s=0x555557b4d600, 
    pkt=0x7fffffffd440) at libavformat/utils.c:1588
#8  0x0000555555bf3012 in avformat_find_stream_info (ic=0x555557b4d600, 
    options=0x555557b87440) at libavformat/utils.c:3813
#9  0x00005555556528a0 in open_input_file (o=0x7fffffffd7a0, 
    filename=0x7fffffffe1c9 "test_complex.avs") at fftools/ffmpeg_opt.c:1127
#10 0x000055555565c706 in open_files (l=0x555557b4d318, 
    inout=0x555556be260f "input", open_file=0x555555651d77 <open_input_file>)
    at fftools/ffmpeg_opt.c:3285
#11 0x000055555565c8be in ffmpeg_parse_options (argc=3, argv=0x7fffffffde08)
    at fftools/ffmpeg_opt.c:3325
--Type <RET> for more, q to quit, c to continue without paging--
#12 0x000055555567c7c8 in main (argc=3, argv=0x7fffffffde08)
    at fftools/ffmpeg.c:4862
(gdb) 
```

magiblot commented:
`Exprfilter::GetFrame` seems to have somehow called a function at address zero.
I suggest you run `cmake` with the argument `-DCMAKE_BUILD_TYPE=Debug` to see source line numbers for libavisynth.so.

qyot27 commented:
```
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff666ff7c in Exprfilter::GetFrame (this=0x555557b82e10, n=0, 
    env=0x555557b78e90) at ../avs_core/filters/exprfilter/exprfilter.cpp:3415
#2  0x00007ffff65e44a1 in Cache::GetFrame (this=0x555557b8c9f0, n=0, 
    env=0x555557b78e90) at ../avs_core/core/cache.cpp:189
#3  0x00007ffff65e2560 in avs_get_frame (p=0x555557b874b0, n=0)
    at ../avs_core/core/avisynth_c.cpp:443
#4  0x0000555555a62cd2 in avisynth_read_packet_video (s=0x555557b4d600, 
    pkt=0x7fffffffd440, discard=0) at libavformat/avisynth.c:671
#5  0x0000555555a63247 in avisynth_read_packet (s=0x555557b4d600, 
    pkt=0x7fffffffd440) at libavformat/avisynth.c:809
#6  0x0000555555be7ea1 in ff_read_packet (s=0x555557b4d600, pkt=0x7fffffffd440)
    at libavformat/utils.c:857
#7  0x0000555555beb0dc in read_frame_internal (s=0x555557b4d600, 
    pkt=0x7fffffffd440) at libavformat/utils.c:1588
#8  0x0000555555bf3012 in avformat_find_stream_info (ic=0x555557b4d600, 
    options=0x555557b87440) at libavformat/utils.c:3813
#9  0x00005555556528a0 in open_input_file (o=0x7fffffffd7a0, 
    filename=0x7fffffffe1d5 "test_complex.avs") at fftools/ffmpeg_opt.c:1127
#10 0x000055555565c706 in open_files (l=0x555557b4d318, 
    inout=0x555556be260f "input", open_file=0x555555651d77 <open_input_file>)
    at fftools/ffmpeg_opt.c:3285
#11 0x000055555565c8be in ffmpeg_parse_options (argc=3, argv=0x7fffffffde08)
--Type <RET> for more, q to quit, c to continue without paging--
    at fftools/ffmpeg_opt.c:3325
#12 0x000055555567c7c8 in main (argc=3, argv=0x7fffffffde08)
    at fftools/ffmpeg.c:4862
(gdb) 
```

magiblot commented:
The initialization of `d.proc[]` in lines 5332 and 5344 seems to be surrounded by `AVS_WINDOWS` and `VS_TARGET_CPU_X86` ifdefs. I don't know if it will work as is on linux.

magiblot commented:
https://github.com/AviSynth/AviSynthPlus/pull/123
There goes dynamic plugin loading 😃:
```
LoadPlugin("../SimpleLUT/simplelut.so")
LUTClip()
```
![imatge](https://user-images.githubusercontent.com/20713561/74581677-434b4200-4fb2-11ea-8df5-5f02faf9b355.png)
```
LoadPlugin("does-not-exist")
```
![imatge](https://user-images.githubusercontent.com/20713561/74581695-7f7ea280-4fb2-11ea-81c9-d97a8ee468d7.png)
```
LoadPlugin("/usr/lib/x86_64-linux-gnu/libdl.so")
```
![imatge](https://user-images.githubusercontent.com/20713561/74581704-a210bb80-4fb2-11ea-9156-9baeca04c16c.png)
But I haven't been able of displaying my filter's output, I don't know why. I can't look into it right now, so maybe you'll be able to try with other filters in the meantime.
```
LoadPlugin("../SimpleLUT/simplelut.so")
LUTClip("YUV", 1) #YUV444P8
```
![imatge](https://user-images.githubusercontent.com/20713561/74581729-f0be5580-4fb2-11ea-833a-c97719a09e36.png)

magiblot commented:
> ![imatge](https://user-images.githubusercontent.com/20713561/74581729-f0be5580-4fb2-11ea-833a-c97719a09e36.png)

This was an issue with FFmpeg that's already solved in the current state of https://github.com/qyot27/FFmpeg/tree/avsplus_linux. So loaded plugins are definitely working.

pinterf commented:
> The linux branch has been rebased against master (for the vfb fix) and force-pushed.

1.) Can you explain this technique? Was it necessary because of concurrent changes occured there which are hard to merge automatically later?
2.) It that date for the first commit (Jan.3, 2014) real?
3.) Theoretically: what can be done about pull requests which contain modifications that are to be be changed after discussing it? How can be achieved that the experimenting phase not eppear in the commit history and only the good solution appear there?
E.g. my parser modification still contains temporary soon-to-be-removed debug and comment lines. Or I can remove them in another "cleanup source" commit later?

Thanks

qyot27 commented:
I'm currently running a high fever, so if this comes out a tad incoherent, I apologize.

Rebasing the fork against current changes means that the entire Linux branch's commits are removed from where it currently branches off, and then re-branched at the current state of the branch it's getting rebased onto.

master -
linux +

How it was before the rebase:
```
------
   \
    ++++++++++++++
```

How it is after the rebase:
```
------
      \
       ++++++++++++++
```

It's not necessarily that it's *hard* to merge them later, but it creates extra merge commits (because if the newer branch has been rebased, it can also be `git cherry-pick`ed or otherwise merely fast-forwarded).  A force-push pushes the newly rebased version of the branch to the repository; it means that anyone tracking it needs to delete the old working branch on their local repo (`git branch -D linux`) and then pull it down again (`git fetch upstream` && `git checkout linux`).  For branches intended for development and which will eventually get applied to the master branch, force-pushing is fine, but you never want to force-push master.

Yes, the date on that commit is real; innocenat was the first one to start work on getting a Linux port going, the rest of the commits were already merged into MT for MinGW/GCC support or otherwise.

Fixing up commits at a later time can be done several ways:
`git rebase -i HEAD~X` where X is the number of commits back to check
In the dialog that comes up you can move the commits around by copy-and-pasting, and when a commit fixing up an old commit is placed right after the old commit, change the label on the new commit from 'commit' to 'fixup'.  Then when git finalizes the changes, it rolls the changes in the new commit into the old commit and keeps the old commit's message (you can change the message or commit itself by using `amend` or `edit` in the `git rebase -i` method).  Another way to work on old commits and re-apply them in a clean state is to `git format-patch -X` (again X = number of commits back from HEAD) and then look at each one or the problematic ones to see where the changes need to go.  Sometimes they'll cleanly reapply, or it gives you a good base from which to split the commit apart manually.

If you have a cleanup commit for those commented out lines, I'll be able to fixup the changes if you'd rather me do it.

pinterf commented:
Thanks for the clarification.
"delete the old working branch on their local repo" So I have to do a clean start for "linux".
Is magiblot's existing pull request against the linux branch affected? Is it still valid either it was before or after the rebase?

qyot27 commented:
magiblot's PR is fine.  When I tested it, I'd done the rebase already.

With Github, when you rebase and force-push like I did, and it affects a pull request, you'll see the list of commits move down in the discussion, instead of being up at the top where they were when the PR was opened.  So the plugin loader fix may have already been done against the rebased branch.

magiblot commented:
I have time and willingness to help, but I have not maintained AviSynth in the past nor used it's interface extensively and it is not clear to me what else has to be done.

Maybe it would help to have a To-Do list with the features that need to be implemented to be on par with Windows, the parts that need to be rewritten and other issues that need to be solved, even new features that cannot work on Windows due to not being backwards-compatible.

qyot27 commented:
The big looming problem is Text Overlay; there are lots of areas that were ifdeffed to AVS_WINDOWS because of things that depended on Text Overlay needing to be disabled in order to build.  As I mentioned farther up, if libass can be used for a _unix variant (or a cross-platform one, if it's performant enough on Windows), that would be ideal.  If not, we may have to go with Pango/Cairo like AvxSynth did, and certainly leave the Windows version alone.

pinterf commented:
What are folks using internal Subtitle for? Isn't it only for not-so-serious e.g. debug purposes?

qyot27 commented:
It's not Subtitle I'm really concerned about, it's Info, Version, Create_MessageClip in general, ApplyMessage functionality.  None of those work without TextOverlay either.  You won't get on-screen debug text from anything.  I guess I was remembering older versions of the patchset where I had to disable a lot more things due to it, but the things that actually are disabled are pretty important in general.

I've gone through all the AVS_WINDOWS checks now, discounted the ones that we don't need to care about or were part of fixes that have already been checked, and this is the result:

**concerns memory constraining; avxsynth solution copied over, but does it need more attention**
	avs_core/core/avisynth.cpp:885:#ifdef AVS_WINDOWS // needs linux alternative

**plugin manager related?**
	avs_core/core/avisynth.cpp:1107:#ifdef AVS_WINDOWS

**probably just for win32, so remove the comment about linux**
	avs_core/core/avisynth.cpp:1112:#ifdef AVS_WINDOWS // needs linux alternative, maybe?

**These have to do with logging; terminal stream output or terminal output colors**
	avs_core/core/avisynth.cpp:1172:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
	avs_core/core/avisynth.cpp:1178:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
	avs_core/core/avisynth.cpp:1184:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
	avs_core/core/avisynth.cpp:1190:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
	avs_core/core/avisynth.cpp:1201:#ifdef AVS_WINDOWS
	avs_core/core/avisynth.cpp:1209:#ifdef AVS_WINDOWS
	avs_core/core/avisynth.cpp:1218:#ifdef AVS_WINDOWS // linux alternative?
	avs_core/core/avisynth.cpp:1239:#ifdef AVS_WINDOWS
	avs_core/core/avisynth.cpp:1248:#ifdef AVS_WINDOWS
	avs_core/core/avisynth.cpp:1252:#ifdef AVS_WINDOWS

**seem to be stragglers from plugin manager**
	avs_core/core/avisynth.cpp:1345:#ifdef AVS_WINDOWS
	avs_core/core/avisynth.cpp:1350:#ifdef AVS_WINDOWS

**getting system memory info**
	avs_core/core/avisynth.cpp:1925:#ifdef AVS_WINDOWS

**ApplyMessage; requires TextOverlay because of font and color manipulation**
	avs_core/core/avisynth.cpp:3093:#ifdef AVS_WINDOWS //fixme

**OpenMP threading in plugins? Got disabled with the script parser fix for linux**
	avs_core/core/avisynth.cpp:3124:#ifdef AVS_WINDOWS

**seh-related, probably win32-only**
	avs_core/core/exception.cpp:120:#ifdef AVS_WINDOWS

**old_working_directory, wchar_t might signal Unicode-specific treatment**
	avs_core/core/internal.h:157:#ifdef AVS_WINDOWS

**Win32-specific Unicode stuff**
	avs_core/core/internal.h:176:#ifdef AVS_WINDOWS

**plugin loader**
	avs_core/core/parser/script.cpp:252:#ifdef AVS_WINDOWS

**more win32 Unicode**
	avs_core/core/parser/script.cpp:283:#ifdef AVS_WINDOWS

**More win32 Unicode**
	avs_core/core/parser/script.cpp:357:#ifdef AVS_WINDOWS

**CWDChanger; if/else copied from AvxSynth**
	avs_core/core/parser/script.cpp:435:#ifdef AVS_WINDOWS

**Even more win32 Unicode, if/elsed, fixed earlier in discussion; needs cleanup of commented lines...do lines 663-666 need to be re-enabled?**
	avs_core/core/parser/script.cpp:568:#ifdef AVS_WINDOWS

**This looks exactly the same?  Why is this here?**
	avs_core/core/parser/script.cpp:699:#ifdef AVS_WINDOWS

**num_audio_samples related; I think the if/else might have been copied from AvxSynth**
	avs_core/core/parser/script.cpp:1420:#ifdef AVS_WINDOWS

**This one was copied from AvxSynth**
	avs_core/core/parser/script.cpp:1469:#ifdef AVS_WINDOWS

**plugin related**
	avs_core/core/parser/script.cpp:1603:#ifdef AVS_WINDOWS

**Windows API checking**
	avs_core/core/parser/script.cpp:1721:#ifdef AVS_WINDOWS

**StrtoUtf8, win32-only**
	avs_core/core/parser/script.cpp:1760:#ifdef AVS_WINDOWS

**plugin-related**
	avs_core/core/parser/script.h:249:#ifdef AVS_WINDOWS

**GetLastErrorText: do we need a Linux equivalent?**
	avs_core/core/PluginManager.cpp:49:#ifdef AVS_WINDOWS // translate to Linux error handling

**AddAutoloadDir**
	avs_core/core/PluginManager.cpp:517:#ifdef AVS_WINDOWS

**plugin search paths, looks like AVS_LINUX ifdef @ 789 is actually solution; turn into if/else?**
	avs_core/core/PluginManager.cpp:758:#ifdef AVS_WINDOWS

**TryAsAvsC**
	avs_core/core/PluginManager.cpp:985:#ifdef AVS_WINDOWS

**_fullpath/realpath if-else**
	avs_core/filters/conditional/conditional_reader.cpp:571:#ifdef AVS_WINDOWS

**_MAX_PATH/PATH_MAX if-else**
	avs_core/filters/conditional/conditional_reader.h:87:#ifdef AVS_WINDOWS

**jitasm-related**
	avs_core/filters/exprfilter/exprfilter.cpp:107:#ifdef AVS_WINDOWS // jitasm currently doesn't build on Linux
	avs_core/filters/exprfilter/exprfilter.cpp:568:#ifdef AVS_WINDOWS // jitasm currently doesn't build on Linux

**if-else needs to be reworked to actually enable the proper d.proc, or get jitasm working so the AVS_WINDOWS isn't necessary**
	avs_core/filters/exprfilter/exprfilter.cpp:5323:#ifdef AVS_WINDOWS

**GetTextBoundingBox is in TextOverlay, Create_MessageClip depends on TA_TOP, TA_CENTER from wingdi**
	avs_core/filters/source.cpp:451:#ifdef AVS_WINDOWS // TA_TOP|CENTER is in windgi, this needs porting

**Version() requires Create_MessageClip**
	avs_core/filters/source.cpp:1963:#ifdef AVS_WINDOWS

**Function exports affected by above**
MessageClip	avs_core/filters/source.cpp:1991:#ifdef AVS_WINDOWS
Version		avs_core/filters/source.cpp:1998:#ifdef AVS_WINDOWS

**Text-Overlay is extremely Windows-specific, and currently toggled on by -DENABLE_FILTER_TEXTOVERLAY on MSVC/MinGW**

qyot27 commented:
@magiblot 
One of the new commits is a fix for GCC 8 to add stdc++fs to `target_link_libraries`.  Does that resolve the issue with needing that library on your end?

magiblot commented:
Yes, it does. Thank you! Thank you too for the detailed summary.

qyot27 commented:
As a general idea, and so long as there's not outstanding bugs inside of the implementation, I think we can go ahead and enable the NEW_AVSVALUE arrays for non-Windows builds.  We're not going to have to worry about plugin compatibility because to get ported to Linux, said plugins will *have* to update their headers and rebuild anew, so there shouldn't be any issue with the backward compatibility problem.  There was only one spot in avisynth.cpp that had to be fixed to let it build on Linux, too (!stricmp->!_stricmp).

If that turns out to be an issue, we can always turn it back off.  I just didn't want to forget about this one before we start finalizing the linux branch (however far away we are from that).

pinterf commented:
I was chasing a framebuffer memory leak today.
It turned out that InterlockedDecrement and InterlockedIncrement simulation in linux.h was returning the original value, while the Windows variant is returning the value _post_ the operation.

pinterf commented:
Well, my last commits did not appear in this pull request automatically, is it limited to 50 commits at a time?

pinterf commented:
I was trying to build for MSVC x64 and in findfirst.cpp there are errors:
`libgen.h` and `dirent.h` is missing.

pinterf commented:
I'm working on a simple Text filter, based on the current info.h, since I missed it in the high bit depth transition, I had to work with it. Though it is using a sinlge fixed font, it can perfectly serve the usual MessageClip/ColorYUV show info/Info(), etc services with a relatively minimal effort.
Now it works like a SubTitle substitution, of course, no font name, size is ignored, etc.
Work in progress.

pinterf commented:
Revive AddAutoloadDir, ClearAutoloadDirs, AutoloadPlugins for linux.
This commit eliminates the need for the findfirst.cpp, I'm using std c++ filesystem support.
findfirst.cpp it is not removed yet though. I've tested on Linux and (with findfirst removed from source) on Windows, and it worked.
- Instead of dll .so is searched on linux 
- File name comparisons are for exact match, since linux is case sensitive
- symlink support is to be decided (see comments in AutoLoadPlugins source)
- AddAutoLoadDir: strings "SCRIPTDIR", "MAINSCRIPTDIR", and "PROGRAMDIR" will be resolved like in Windows.

```
AddAutoLoadDir(".")
AddAutoLoadDir("/lib")

```
I noticed that when non-avs plugin .so files are found, the script will exit instead of ignoring them (?)

note: call for FunctionExists will always invoke AutoLoadPlugins

pinterf commented:
> 
> 
> I was trying to build for MSVC x64 and in findfirst.cpp there are errors:
> `libgen.h` and `dirent.h` is missing.

This is probably solved by not using findfirst cpp at all.

qyot27 commented:
Symlinking will allow for one scenario at least: a normal library that happens to have an AviSynth plugin interface can be symlinked into the autoload plugin directory.  This is precisely how it works for VapourSynth and AvxSynth regarding FFMS2 (which gets installed to `<prefix>/lib`, while the `<prefix>/lib/avxsynth` and `<prefix>/lib/vapoursynth` directories have symlinks to it rather than duplicate copies).

pinterf commented:
Arghh. When you use VS2019 IDE and editor keeps inserting tabs instead of spaces _despite_ the Text Editor > C/C++ > Tabs settings, then navigate to menu Tools > Options, then Text Editor > Advanced and _uncheck_ "Use adaptive formatting" .
(Coding convention is to use spaces in Avisynth+)

pinterf commented:
Text filter is still under construction.
It's SubTitle-like but only a single fixed font, no font angles, but we have textcolor, outline with halo color, alignment, utf8, multiline. Info(), the internal ApplyMessage from e.g. ColorYUV, ShowFrameNumber are working fine but there is still mess in the source. RGB24/32/YUY2 along with RGB48/64 support will come later when the planar formats are really finished.
Till then here is a screenshot
https://drive.google.com/open?id=160SYDsg3Mh5AvL3ON6HNOhA7VUbycVpG

qyot27 commented:
Effectively, the list I posted earlier was just a cursory points list of wherever AVS_WINDOWS showed up; the ones I'd consider to be critical are much fewer:

**These have to do with logging; terminal stream output or terminal output colors**
avs_core/core/avisynth.cpp:1172:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
avs_core/core/avisynth.cpp:1178:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
avs_core/core/avisynth.cpp:1184:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
avs_core/core/avisynth.cpp:1190:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
avs_core/core/avisynth.cpp:1201:#ifdef AVS_WINDOWS
avs_core/core/avisynth.cpp:1209:#ifdef AVS_WINDOWS
avs_core/core/avisynth.cpp:1218:#ifdef AVS_WINDOWS // linux alternative?
avs_core/core/avisynth.cpp:1239:#ifdef AVS_WINDOWS
avs_core/core/avisynth.cpp:1248:#ifdef AVS_WINDOWS
avs_core/core/avisynth.cpp:1252:#ifdef AVS_WINDOWS

**getting system memory info**
avs_core/core/avisynth.cpp:1925:#ifdef AVS_WINDOWS

**Even more win32 Unicode, if/elsed, fixed earlier in discussion; needs cleanup of commented lines...do lines 663-666 need to be re-enabled?**
avs_core/core/parser/script.cpp:568:#ifdef AVS_WINDOWS

These are all the related to the text stuff that pinterf is working on:
```
ApplyMessage; requires TextOverlay because of font and color manipulation
avs_core/core/avisynth.cpp:3093:#ifdef AVS_WINDOWS //fixme

GetTextBoundingBox is in TextOverlay, Create_MessageClip depends on TA_TOP, TA_CENTER from wingdi
avs_core/filters/source.cpp:451:#ifdef AVS_WINDOWS // TA_TOP|CENTER is in windgi, this needs porting

Version() requires Create_MessageClip
avs_core/filters/source.cpp:1963:#ifdef AVS_WINDOWS

Function exports affected by above
MessageClip avs_core/filters/source.cpp:1991:#ifdef AVS_WINDOWS
Version avs_core/filters/source.cpp:1998:#ifdef AVS_WINDOWS

Text-Overlay is extremely Windows-specific, and currently toggled on by -DENABLE_FILTER_TEXTOVERLAY on MSVC/MinGW
```


So, 3 things (which may not be very critical, just annoying), 4 if we include the bundle of text-related fixes.  If need be, the text filter could be something that's fixed up after the release of 3.5, because it doesn't impede the ability to load video and filter it in most ways.  The remaining AVS_WINDOWS blocks are either cases where what was broken was already fixed, or are minor (given that AviSynth+ is already in a state where it runs) and could be refactored away in the future.

There's also the requirement of using `-fpermissive`.  That really shouldn't be necessary to use, since we don't have to use it on Windows.

One of the other critical points concerning getting on equal footing with AvxSynth is that it also supported OSX and BSD.  I have occasional access to a Mac, although it's currently running High Sierra.  I'm looking at setting up a FreeBSD VM in VirtualBox to test the BSD side, and given that A) macOS is a BSD derivate and B) there's nothing that should be using the Apple-specific frameworks, it's probably going to be fine and require only minor tweaking (AvxSynth only required small bits to get it to work on them, part of it was already merged on our side, and the other part was in their plugin manager, IIRC).  The real limiting factor will probably be which version of Clang or GCC (via Homebrew, most likely; I'm sure the Apple Developer Tools only use Clang nowadays) is available and how to handle that.

magiblot commented:
> **These have to do with logging; terminal stream output or terminal output colors**
> avs_core/core/avisynth.cpp:1172:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
> avs_core/core/avisynth.cpp:1178:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
> avs_core/core/avisynth.cpp:1184:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
> avs_core/core/avisynth.cpp:1190:#ifdef AVS_WINDOWS // FOREGROUND_* is Windows-specific
> avs_core/core/avisynth.cpp:1201:#ifdef AVS_WINDOWS
> avs_core/core/avisynth.cpp:1209:#ifdef AVS_WINDOWS
> avs_core/core/avisynth.cpp:1218:#ifdef AVS_WINDOWS // linux alternative?
> avs_core/core/avisynth.cpp:1239:#ifdef AVS_WINDOWS
> avs_core/core/avisynth.cpp:1248:#ifdef AVS_WINDOWS
> avs_core/core/avisynth.cpp:1252:#ifdef AVS_WINDOWS

I'm not suggesting this feature should be removed forever, but terminal formatting could be probably disabled for now or left as is, unimplemented on linux.

The implementation details are: Windows has a Console API which allows you to easily format your terminal output. Everyone else uses [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters), which are also supported by the Windows console in recent versions. The problem with ANSI escape codes is that they don't look right when redirected to a file or when read by another application. So, using them comes at the price of having to determine whether stdout is a terminal or not (which is what POSIX's `isatty` precisely does on both [linux](http://man7.org/linux/man-pages/man3/isatty.3.html) and [Windows](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/posix-isatty)).

Implementing this strategy is not hard at all but it's not worth making our code complex, unless we come up with a simple interface. In addition, AviSynth is not a terminal application and these messages are only useful for debugging purposes. Since the logging messages already begin with `ERROR`, `WARNING` and the like, they already serve their purpose well.

qyot27 commented:
Okay, I was not entirely sure whether it was just the colorization that was the issue there or if it was a deeper issue with the logging function.  Knowing it's just the color output means that no, I don't think that's critical at all.

And on the macOS/BSD front, I did manage to get AviSynth+ and FFmpeg+patch compiled on FreeBSD 12 and verified that Colorbars works.  There were some things that needed changing, but on the whole it's fine...so long as GCC is the compiler.  By default, FreeBSD uses Clang (version 8 in FreeBSD 12), and that was giving me all kinds of trouble because our current Clang support is limited to clang-cl under MSVC.  Getting it to build with GCC on FreeBSD was just a matter of cleaning up the OS detection defines and if-else-ing the couple of places where we include `malloc.h`.

So I'd expect macOS to be pretty similar, except that getting it to build under Clang in its normal GCC-compatible mode needs to be done at some point so there's not a big caveat.

pinterf commented:
First:
Under Linux the only valid script source encoding is utf8.
I wish it would apply on Windows version as well.

- SimpleText filter with the same parameters as SubTitle, size, orientatio, font parameters are not working, we have a single fixed 10x20 matrix character set. with limited character set: ASCII and Latin1, from the original one.
- Character definition matrix in info.h is extended to have a proper unicode codepoint, I even added some missing characters from my language.

Over the old info.h debug text there are further enhancements
- ansi and utf8 conversion, internally handled as utf16
- background fading of original code is kept, when MSB of halocolor is $FF (e.g. $FF000000)
- color
- halocolor (outline), no outline when MSB of halocolor is not $00
- aligment (H and V)
- multiline

- Subtitle is redirected to SimpleText

For build, please enable back ENABLE_FILTER_TEXTOVERLAY in CMakeList, I have not commited that change, maybe this option can be deleted at the moment.

pinterf commented:
I have noticed that AVX2 was not shown in Info() under Linux. We have to check it, why.

qyot27 commented:
Pushed a few cleanups (findfirst.cpp removal and the comments in script.cpp), re-enabled Text Overlay at all times, formatted the filter lists in avisynth.cpp in a clearer form, and the BSD support commit (the Clang stuff in there probably needs fixing, it was leftover from experimenting with that before I gave up and used GCC to confirm it worked).

Technically, AVS_LINUX could probably be abstracted into AVS_POSIX or something that's true for all the non-Windows OSes, and the use of AVS_LINUX could be reserved for anything that really really needs something specific to Linux.  Because it's a little weird that BSD requires AVS_LINUX and AVS_BSD to both be true.  That's mostly a bikeshedding thing, though, and in no way critical.  As long as it works, it can just be a little weird.

Still need to test with macOS, and the NEWVALUE arrays need to be enabled for Linux in a smarter way than just enabling them through CMakeLists.txt.

qyot27 commented:
Also, `-fpermissive` is still required to get around this error:
```
[ 24%] Building CXX object avs_core/CMakeFiles/AvsCore.dir/core/avisynth.cpp.o
/home/qyot27/AviSynthPlus/avs_core/core/avisynth.cpp: In member function ‘virtual char* ScriptEnvironment::VSprintf(const char*, void*)’:
/home/qyot27/AviSynthPlus/avs_core/core/avisynth.cpp:3119:50: error: ISO C++ forbids casting to an array type ‘va_list’ {aka ‘__va_list_tag [1]’} [-fpermissive]
 3119 |     std::string str = FormatString(fmt, (va_list)val);
      |                                                  ^~~
make[2]: *** [avs_core/CMakeFiles/AvsCore.dir/build.make:297: avs_core/CMakeFiles/AvsCore.dir/core/avisynth.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:156: avs_core/CMakeFiles/AvsCore.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

But that looks to be the only place.

pinterf commented:
That va_list issue on x64 vs x86, compiler1 vs compiler2 has great literature.
For the record, some articles on the topic:
https://stackoverflow.com/questions/26918450/how-is-vprintf-implemented/26919307#26919307
https://stackoverflow.com/questions/8047362/is-gcc-mishandling-a-pointer-to-a-va-list-passed-to-a-function/8048892#8048892
https://stackoverflow.com/questions/4958384/what-is-the-format-of-the-x86-64-va-list-structure/4958507#4958507

And as the comment in avisynth.h says:
```
  // note: val is really a va_list; I hope everyone typedefs va_list to a pointer
  virtual char* __stdcall VSprintf(const char* fmt, void* val) = 0;
```
This "Let's hope" is not a strong reasoning :)

I wonder why (void *) was chosen instead of the logical va_list?

qyot27 commented:
Upon further testing, macOS is now working too.  Like BSD, it requires using GCC, not Clang, and in most cases, it's changing the `malloc.h` if-elses to detect AVS_BSD or AVS_MACOS, but there's also a necessary change in CMakeLists because the linker¹ on macOS needs to have `-Wl,-undefined,error` instead of `-Wl,--no-undefined`.

Also, `uchar.h` doesn't exist on macOS, but doesn't seem to cause issues if I just disable it with `#if 0`.  It probably doesn't exist on BSD, either (but I tested BSD before the text filter fix, so I'll have to fix up the VM again for that).

¹I couldn't seem to get a clear answer on whether that was `lld` or more likely, the one from Apple's `binutils` package.  It didn't require that on BSD, at any rate.

qyot27 commented:
I managed to get a VM of macOS Mojave set up on a computer I have considerably better access to, and ran some more tests.

AddAutoloadDir fails on Mac, results in `Evaluate: Unhandled C++ exception!`, but manually using LoadPlugin works as expected.  This possibly also affects BSD, but I haven't checked with that yet.

pinterf commented:
> 
> I managed to get a VM of macOS Mojave set up on a computer I have considerably better access to, and ran some more tests.
> 
> AddAutoloadDir fails on Mac, results in `Evaluate: Unhandled C++ exception!`, but manually using LoadPlugin works as expected. This possibly also affects BSD, but I haven't checked with that yet.

What files did that autoload directory contain? I mentioned erlier that if a non-avisynth .so file exist there then error will be thrown.

qyot27 commented:
No, it wasn't that.  The only thing in /usr/local/lib/avisynth was a symlink to the dylib for ffms2 (I do need to adjust the spot in PluginManager where I added *.dylib as the extension for macOS, but the C++ exception was thrown in a build that had that corrected).

qyot27 commented:
I have force-pushed the linux branch again due to some cleanups and commit message corrections.  I've also updated the FFmpeg patch with the updated headers (posix.h and macOS/BSD support) and force-pushed that branch on my own repo as well.

qyot27 commented:
Okay, I think we're ready for merging this back to master and making an official 3.5 release.

The AddAutoloadDir problem on macOS/BSD and that lone `-fpermissive` that's still required do need to be fixed, but because these **only** impact the non-Windows builds, I think it'd be fine to let those end up being in 3.5.1, especially since non-Windows users will likely be fine just grabbing from git to get those fixed as they occur, rather than being super-concerned with the release tarballs.  Same with getting Clang working in non-MSVC-compat mode.  That's a big thing that needs fixing, but since it will compile and work if built with GCC, that's mostly for convenience.

I'll also make a separate forum thread on Doom9 to cover the Linux/macOS/BSD side, because as it's a really new feature, users testing and reporting bugs should probably be more focused than the general AviSynth+ thread.

qyot27 commented:
3.5 release made, threads posted on Doom9, and email about AvxSynth->AviSynth+ transition sent to the ffmpeg-devel mailing list (although I'm not seeing it appear in the Archives yet).

pinterf commented:
Thanks, I see you have noticed my unsynchronized commit.

qyot27 commented:
Since this has been merged and general bug reports and whatnot are being discussed in their own threads or over on Doom9, I'll go ahead and close this now.
